src/modm/processing/fiber/scheduler.hpp.in

```diff
@@ -0,0 +1,181 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include "fiber.hpp"
+%% if multicore
+#include <modm/platform/core/multicore.hpp>
+%% for i in range(num_cores)
+#ifndef MODM_FIBER_STORAGE_{{i}}
+#define MODM_FIBER_STORAGE_{{i}} MODM_MULTICORE_STORAGE_{{0}}(data.fiber)
+#endif
+%% endfor
+%% else
+#ifndef MODM_FIBER_STORAGE
+#define MODM_FIBER_STORAGE
+#endif
+%% endif
+
+namespace modm::fiber
+{
+
+void yield();
+
+class Scheduler
+{
+	friend class ::modm::Fiber;
+	friend class Waitable;
+	friend void yield();
+	Scheduler(const Scheduler&) = delete;
+	Scheduler() = delete;
+
+protected:
+	struct Data {
+		/// Last fiber in the ready queue.
+		Fiber* last = nullptr;
+		/// Current running fiber
+		Fiber* current = nullptr;
+		inline void registerFiber(Fiber* fiber) {
+			if (last == nullptr)
+			{
+				fiber->next = fiber;
+				last = fiber;
+				return;
+			}
+			runLast(fiber);
+		}
+		inline void runLast(Fiber* fiber)
+		{
+			fiber->next = last->next;
+			last->next = fiber;
+			last = fiber;
+		}
+		inline Fiber* removeCurrent() {
+			if (current == last) last = nullptr;
+			else last->next = current->next;
+			current->next = nullptr;
+			return current;
+		}
+		inline bool empty() const {
+			return last == nullptr;
+		}
+		inline void jump(Fiber& other) {
+			auto from = current;
+			current = &other;
+			modm_context_jump(&(from->ctx.sp), other.ctx.sp);
+		}
+	};
+%% if multicore
+	%% for i in range(num_cores)
+		MODM_FIBER_STORAGE_{{i}} static inline  Data data{{i}} = {nullptr,nullptr};
+	%% endfor
+	static constexpr Data* lookup[] = {
+%% for i in range(num_cores)
+			&data{{i}},
+%% endfor
+	};
+	static inline Data& getData() {
+		return *lookup[multicore::Core::getNum()];
+	}
+%% else
+	MODM_FIBER_STORAGE static inline Data data = {nullptr,nullptr};
+	static inline Data& getData() { return data; }
+%% endif
+	static inline void jump(Fiber& from,Fiber& to) {
+		auto& d = getData();
+		d.current = &to;
+		modm_context_jump(&(from.ctx.sp), to.ctx.sp);
+	}
+public:
+	// Should be called by the main() function.
+	static inline bool
+	run()
+	{
+		auto& d = getData();
+		if (d.last == nullptr) return false;
+		d.current = d.last->next;
+		modm_context_start(d.current->ctx.sp);
+		return true;
+	}
+
+	static inline bool
+	empty()
+	{
+		return getData().empty();
+	}
+
+	static inline Fiber*
+	removeCurrent()
+	{
+		return getData().removeCurrent();
+	}
+
+	static inline void
+	runNext(Fiber* fiber)
+	{
+		auto& d = getData();
+		fiber->next = d.current->next;
+		d.current->next = fiber;
+	}
+
+	static inline void
+	runLast(Fiber* fiber)
+	{
+		getData().runLast(fiber);
+	}
+
+protected:
+%% if multicore
+	static inline void
+	registerFiber(Fiber* fiber,size_t core)
+	{
+		if (core != multicore::Core::getNum())
+		{
+			multicore::SystemSpinLockGuard g;
+			lookup[core]->registerFiber(fiber);
+			return;
```

Comment:
Yes, but shouldn't the `multicore::SystemSpinLockGuard g;` be used in *every* function that modifies the `Data` structure? What if CPU0 registers a fiber on it's own core while CPU1 registers a fiber on CPU0 core?

Reply:
Yes, you right. At this point need protect on any call, all other functions called only on one core.